lib/archethic/self_repair/network_view.ex

```diff
@@ -0,0 +1,186 @@
+defmodule Archethic.SelfRepair.NetworkView do
+  @moduledoc """
+  The network view is 2 things:
+  - the P2P view (list of available & authorized nodes)
+  - the Network chains view (oracle/origin/nodesharedsecrets)
+
+  It is useful to compare with other nodes to detect desynchronization.
+  The P2P view is handled by the P2P module, we just do the hash here for convenience.
+  """
+
+  use GenServer
+  @vsn Mix.Project.config()[:version]
+
+  alias Archethic.OracleChain
+  alias Archethic.Crypto
+  alias Archethic.P2P
+  alias Archethic.PubSub
+  alias Archethic.SharedSecrets
+  alias Archethic.TransactionChain
+  alias Archethic.TransactionChain.Transaction
+
+  require Logger
+
+  defmodule State do
+    defstruct [
+      # binary()
+      node_shared_secrets: nil,
+      # binary()
+      oracle: nil,
+      # list(binary())
+      origin: []
+    ]
+  end
+
+  #               _
+  #    __ _ _ __ (_)
+  #   / _` | '_ \| |
+  #  | (_| | |_) | |
+  #   \__,_| .__/|_|
+  #        |_|
+
+  @doc """
+  Start the NetworkView server
+  """
+  @spec start_link(list()) :: GenServer.on_start()
+  def start_link([]) do
+    GenServer.start_link(__MODULE__, [], name: __MODULE__)
+  end
+
+  @doc """
+  Return the hash of the P2P view
+  """
+  @spec get_p2p_hash() :: binary()
+  def get_p2p_hash() do
+    # this is not using the genserver
+    # since the nodes' state is already in the P2P module
+    P2P.authorized_and_available_nodes()
+    |> Enum.map_join(& &1.last_public_key)
+    |> Crypto.hash()
+  end
+
+  @doc """
+  Return the hash of the network chains view
+  """
+  @spec get_chains_hash() :: binary()
+  def get_chains_hash() do
+    GenServer.call(__MODULE__, :get_chains_hash)
+  end
+
+  @doc """
+  Update the state with given transaction.
+  GenServer is called only on relevant transactions.
+  """
+  @spec load_transaction(Transaction.t()) :: :ok
+  def load_transaction(%Transaction{
+        type: type,
+        address: address
+      })
+      when type in [:node_shared_secrets, :oracle, :origin] do
+    GenServer.cast(__MODULE__, {:load_transaction, type, address})
+  end
+
+  def load_transaction(_), do: :ok
+
+  #            _ _ _                _
+  #   ___ __ _| | | |__   __ _  ___| | _____
+  #  / __/ _` | | | '_ \ / _` |/ __| |/ / __|
+  # | (_| (_| | | | |_) | (_| | (__|   <\__ \
+  #  \___\__,_|_|_|_.__/ \__,_|\___|_|\_|___/
+  #
+
+  def init([]) do
+    state =
+      if Archethic.up?() do
+        fetch_initial_state()
+      else
+        Logger.info("NetworkView: Waiting for Node to complete Bootstrap. ")
+        PubSub.register_to_node_status()
+        :not_initialized
+      end
+
+    {:ok, state}
+  end
+
+  def handle_call(:get_chains_hash, _from, state = %State{}) do
+    hash =
```

Comment:
Why don't you prepare the hash once and cached it, until a new transaction invalid it and needs to be recomputed.
This function is called for each transaction, I'm afraid of the bottleneck. Ideally, an ETS would solve the bottlenecks for this high-frequency call.